File: Finabel.java
import java.math.BigInteger;

public class Finabel {

    static String hexadecimal = "0123456789abcdef";
    static String lookupHex[] = new String[256];
    static int lookupCode[] = new int[256];
    static String record_separator = null;
    static String field_separator = null;
    static boolean uninitialized = true;

    Finabel() {
        if (uninitialized) {
            uninitialized = false;

            /*
 Static lookup tables
*/
            for (int index = 0; index < 256; ++index)
                lookupHex[index] = String.valueOf(hexadecimal.charAt(index >> 4)) +
                String.valueOf(hexadecimal.charAt(index & 0xf));

            for (int index = 0x30; index < 0x3a; ++index)
                lookupCode[index] = index - 0x30;
            for (int index = 0x61; index < 0x67; ++index)
                lookupCode[index] = index - 0x57;

            /*
 Separators improve immalleability 
*/

            record_separator = toHex("\u001e");
            field_separator = toHex("\u001c");
        }
    }

    /*
 Convert a utf-8 string to hexadecimal digits
*/

    private static String toHex(String text) {
        String result = "";
        byte encoded[] = null;
        try {
            encoded = text.getBytes("UTF8");
        } catch (java.io.UnsupportedEncodingException error) {}
        for (int index = 0; index < encoded.length; ++index)
            result += lookupHex[encoded[index] & 0xff];
        return result;
    }

    /*
 A few large safe primes
*/

    static BigInteger A = new BigInteger(
        "90086843365112375319585743415488659286349207571888" +
        "72143870468294068098052833612962771759906636851615" +
        "30183997243896077623165157756007099732429029873106" +
        "25906982188676619566197948156310182642979757089086" +
        "64735135318987857748964189266150597208152376641168" +
        "12063491035355207065882456370964859448027182804663"
    );

    static BigInteger B = new BigInteger(
        "90674648893112227992308781677425426254932596102120" +
        "36868710245939602733644836408374612756513538399124" +
        "78276312989231907136777434371442297783809058139703" +
        "70454811258254923105788711983760996127549469859402" +
        "84644249381392368127863251805736853657160545788782" +
        "69276609138278646089652565880503149299047623725983"
    );

    static BigInteger C = new BigInteger(
        "93596367994515962161810813565073160231612346284473" +
        "99189667910540022206214547335159626318385581670717" +
        "14943415781502503512108093455147689164719674990397" +
        "03576424880848675456223672701325547389408057502297" +
        "15406770374497502730147945284076674546501315764540" +
        "15775014701175216242011377646611112897139737772263"
    );

    static int minimum_digits = C.toString(16).length();

    /*
 Key stretching function
*/

    private static BigInteger stretch(BigInteger value) {
        String hexadecimal = value.toString(16);
        String buffer = hexadecimal;
        do {
            buffer += (record_separator + hexadecimal);
        } while (buffer.length() < minimum_digits);
        return new BigInteger(buffer, 16);
    }

    /*
 Cycle once through our finite fields
*/

    private static BigInteger cycle(BigInteger V) {
        BigInteger Q = stretch(V);
        BigInteger R = Q.multiply(A).mod(B);
        BigInteger S = stretch(R);
        return Q.multiply(S).mod(C);
    }

    /*
 Hash function interface
*/

    public String hash(String keys[], int rounds, int digits, int cost) {

        if (rounds == 0)
            rounds = 1024;
        if (cost == 0)
            cost = 512;
        cost *= 1024;
        /*
 Initial construction: concatenate keys/salt
*/
        String merged = record_separator;
        for (int index = 0, limit = keys.length; index < limit; ++index) {
            String next = keys[index];
            if (next == null || next == "") continue;
            merged += toHex(next) + field_separator;
        }

        BigInteger V = new BigInteger(merged, 16);

        /*
 Estimate the number of rounds needed to construct the result
*/

        int window = cost / minimum_digits + 1;
        int discards = (window <= rounds) ? rounds - window : 0;

        /*
 Spin through "discard" rounds   
*/

        while (discards-- > 0) V = cycle(V);

        /*
 Finish off with enough rounds needed satisfy our memory quota   
*/

        String buffer = "";
        while (true) {
            V = cycle(V);
            buffer += V.toString(16);
            if (buffer.length() >= cost)
                break;
        }

        /*
  Build the result using memory-dependant construction, back to front
*/
        String result = "";
        int current = buffer.length() - 1;
        while(true) {
            int read = current;
            int accumulator = 0;

            while (accumulator < window) {
                if (read-- == 0) break;
                accumulator <<= 4;
                accumulator |= lookupCode[buffer.codePointAt(read)];
            }
            int offset = (int)(Math.floor(accumulator % window) + 1);
            if (offset >= current) break;
            current -= offset;
            result += buffer.charAt(current);
        }

        /*
  Truncate or pad the result, if necessary
*/
        if (digits > 0) {
            int length = result.length();
            if (length > digits) return result.substring(0, digits);
            while (length++ < digits) result = result.concat("0");
        }

        return result;
    }

    /*
  Example usage
*/

    public static void main(String[] argv) {

        int argc = argv.length;
        if (argc <= 1)
            System.out.println("Usage: java Finabel.java [KEY] [SALT] [ROUNDS] [DIGITS] [COST]");
        String key = "";
        if (argc > 0)
            key = argv[0];
        String salt = "";
        if (argc > 1)
            salt = argv[1];
        int rounds = 0;
        if (argc > 2)
            rounds = Integer.parseInt(argv[2]);
        int digits = 0;
        if (argc > 3)
            digits = Integer.parseInt(argv[3]);
        int cost = 0;
        if (argc > 4)
            cost = Integer.parseInt(argv[4]);
        String keys[] = {
            key,
            salt
        };

        Finabel finabel = new Finabel();
        System.out.println(finabel.hash(keys, rounds, digits, cost));
    }
}

File: finabel.py
import math

# Static lookup tables

lookupHex = [None] * 256
lookupCode = [None] * 256
uninitialized = True

# Convert a utf-8 string to hexadecimal digits

def toHex(text):
    global lookupHex
    result = ''
    bytes = text.encode()
    for index in range(len(bytes)):
        result += lookupHex[bytes[index]]
    return result

# A few large safe primes

A = 900868433651123753195857434154886592863492075718887214387046829406809805283361296277175990663685161530183997243896077623165157756007099732429029873106259069821886766195661979481563101826429797570890866473513531898785774896418926615059720815237664116812063491035355207065882456370964859448027182804663

B = 906746488931122279923087816774254262549325961021203686871024593960273364483640837461275651353839912478276312989231907136777434371442297783809058139703704548112582549231057887119837609961275494698594028464424938139236812786325180573685365716054578878269276609138278646089652565880503149299047623725983

C = 935963679945159621618108135650731602316123462844739918966791054002220621454733515962631838558167071714943415781502503512108093455147689164719674990397035764248808486754562236727013255473894080575022971540677037449750273014794528407667454650131576454015775014701175216242011377646611112897139737772263

minimum_digits = len(hex(C)[2:])

# Separators improve immalleability

record_separator = None
field_separator = None

# Key stretching function

def stretch(value):
    global record_separator, minimum_digits
    hexadecimal = hex(value)[2:]
    buffer = hexadecimal
    while True:
        buffer += record_separator + hexadecimal
        if len(buffer) >= minimum_digits:
            break
    return int(buffer, 16)

# Cycle once through our finite fields

def cycle(V):
    global A, B, C
    Q = stretch(V)
    R = Q * A % B
    S = stretch(R)
    return Q * S % C

# Hash function interface

def finabel(
    key,
    salt,
    rounds,
    digits,
    cost,
    ):
    global uninitialized, lookupHex, lookupCode, record_separator, \
        field_separator, minimum_digits
    if uninitialized:
        uninitialized = False
        hexadecimal = '0123456789abcdef'
        for index in range(256):
            lookupHex[index] = hexadecimal[index >> 4] \
                + hexadecimal[index & 0xF]
        for index in range(0x30, 0x3a):
            lookupCode[index] = index - 0x30
        for index in range(0x61, 0x67):
            lookupCode[index] = index - 0x57
        record_separator = toHex("\u001e")
        field_separator = toHex("\u001c")
    if rounds is None or rounds == 0:
        rounds = 1024
    if cost is None or cost == 0:
        cost = 512
    cost *= 1024
    keys = (key if isinstance(key, list) else [key])
    keys.append(salt)

# Initial construction: concatenate keys/salt

    merged = record_separator
    for index in range(len(keys)):
        next = keys[index]
        if next is None or next == '':
            continue
        merged += toHex(next) + field_separator
    V = int(merged, 16)

# Estimate the number of rounds needed to construct the result

    window = math.floor(cost / minimum_digits) + 1
    discards = (rounds - window if window <= rounds else 0)

# Spin through "discard" rounds

    while discards > 0:
        discards -= 1
        V = cycle(V)

# Finish off with enough rounds needed satisfy our memory quota

    buffer = ''
    while True:
        V = cycle(V)
        buffer += hex(V)[2:]
        if len(buffer) >= cost:
            break

# Build the result using memory-dependant construction, back to front

    result = ''
    current = len(buffer) - 1
    while True:
        read = current
        accumulator = 0
        while accumulator < window:
            if read == 0:
                break
            read -= 1        
            accumulator <<= 4
            accumulator |= lookupCode[buffer[read].encode()[0]]
        offset = math.floor(accumulator % window) + 1
        if offset >= current:
            break
        current -= offset
        result += buffer[current]

# Truncate or pad the result, if necessary

    if digits > 0:
        length = len(result)
        if length > digits:
            return result[0:digits]
        while length < digits:
            result += '0'
            length += 1
    return result

